homeassistant/components/zha/api.py

```diff
@@ -0,0 +1,420 @@
+SCHEMA_WS_CLUSTER_COMMANDS = websocket_api.BASE_COMMAND_MESSAGE_SCHEMA.extend({
+    vol.Required(TYPE): WS_ENTITY_CLUSTER_COMMANDS,
+    vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+    vol.Required(ATTR_IEEE): str,
+    vol.Required(ATTR_CLUSTER_ID): int,
+    vol.Required(ATTR_CLUSTER_TYPE): str
+})
+
+
+async def async_load_api(hass, application_controller, listener):
+    """Set up the web socket API."""
+    async def permit(service):
+        """Allow devices to join this network."""
+        duration = service.data.get(ATTR_DURATION)
+        _LOGGER.info("Permitting joins for %ss", duration)
+        await application_controller.permit(duration)
+
+    hass.services.async_register(DOMAIN, SERVICE_PERMIT, permit,
+                                 schema=SERVICE_SCHEMAS[SERVICE_PERMIT])
+
+    async def remove(service):
+        """Remove a node from the network."""
+        from bellows.types import EmberEUI64, uint8_t
+        ieee = service.data.get(ATTR_IEEE_ADDRESS)
+        ieee = EmberEUI64([uint8_t(p, base=16) for p in ieee.split(':')])
+        _LOGGER.info("Removing node %s", ieee)
+        await application_controller.remove(ieee)
+
+    hass.services.async_register(DOMAIN, SERVICE_REMOVE, remove,
+                                 schema=SERVICE_SCHEMAS[IEEE_SERVICE])
+
+    async def reconfigure_device(service):
+        """Reconfigure a ZHA device entity by its ieee address."""
+        ieee = service.data.get(ATTR_IEEE_ADDRESS)
+        entities = listener.get_entities_for_ieee(ieee)
+        _LOGGER.debug("Reconfiguring device with ieee_address: %s", ieee)
+        for entity in entities:
+            if hasattr(entity, 'async_configure'):
+                hass.async_create_task(entity.async_configure())
+
+    hass.services.async_register(DOMAIN, SERVICE_RECONFIGURE_DEVICE,
+                                 reconfigure_device, schema=SERVICE_SCHEMAS[
+                                     IEEE_SERVICE
+                                 ])
+
+    async def set_zigbee_cluster_attributes(service):
+        """Set zigbee attribute for cluster on zha entity."""
+        entity_id = service.data.get(ATTR_ENTITY_ID)
+        cluster_id = service.data.get(ATTR_CLUSTER_ID)
+        cluster_type = service.data.get(ATTR_CLUSTER_TYPE)
+        attribute = service.data.get(ATTR_ATTRIBUTE)
+        value = service.data.get(ATTR_VALUE)
+        manufacturer = await _get_manufacturer(service.data)
+        component = hass.data.get(entity_id.split('.')[0])
+        entity = component.get_entity(entity_id)
+        response = None
+        if entity is not None:
+            response = await entity.write_zigbe_attribute(
+                cluster_id,
+                attribute,
+                value,
+                cluster_type=cluster_type,
+                manufacturer=manufacturer
+            )
+        _LOGGER.debug("Set attribute for: %s %s %s %s %s %s %s",
+                      "{}: [{}]".format(ATTR_CLUSTER_ID, cluster_id),
```

Comment:
Shall we format `cluster_id` as zero padded hex? Since in ZCL specs attrs are all referenced using hex IDs

Reply:
I thought about this... but most of the logging prints them out normally. I think we can take a pass in a subsequent PR and convert all of it.